changes/2023-09-26_encryption-context-decrypt/background.md

```diff
@@ -0,0 +1,584 @@
+[//]: # "Copyright Amazon.com Inc. or its affiliates. All Rights Reserved."
+[//]: # "SPDX-License-Identifier: CC-BY-SA-4.0"
+
+# Which Encryption Context should be used in Decrypt Operations
+
+## Definitions
+
+### Conventions used in this document
+
+The key words
+"MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL"
+in this document are to be interpreted as described in
+[RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+## Issues and Alternatives
+
+### What kind of Encryption Context should Decrypt return?
+
+Decrypt returns Encryption Context.
+This requirement may be satisfied by returning the parsed header that contains this value.
+
+With the addition of the [required encryption context cmm](../../framework/required-encryption-context-cmm.md),
+the encrypt API is able to filter out encryption context key-value pairs that
+are not stored on the message.
+
+Currently, Decrypt takes in optional encryption context.
+This optional encryption context is not stored in the message header, but is still authenticated.
+
+However, the CMM's DecryptMaterials is allowed to modify the encryption context on decrypt.
+A poorly designed CMM could theoretically return materials with an "incorrect" encryption context that contradicts the encryption context in the header of the message.
+
+It MUST be required that the decryption materials obtained from the underlying CMM's MUST
+contain all configured encryption context keys in its encryption context and the values MUST remain
+unchanged.
+
+The Decrypt API MUST return the set of the union of the encryption context stored in the header and the encryption
+context used for authentication only.
+
+### What changes to the existing Required Encryption Context CMM are required?
+
+None
+
+### API changes
+
+If any required encryption context keys from the CMM's `DecryptionMaterials` response
+match any keys stored in the message header, their values MUST match.
+
+```
+forall k <- decryptionMaterials.requiredEncryptionContextKeys
+   | k in headerEncryptionContext
+   :: decryptionMaterials[k] == headerEncryptionContext[k]
+```
+
+## Testing
```

Comment:
Thinking more about this, I think it would be better to more clearly define what the matrix is we care about, as opposed to listing out the different tests and code for them. I would prefer we cut the example code, but make it extremely clear what every case we need to check is in a way that makes it easy to reason that it is complete. For example, maybe frame it like this:

---

First, suppose we have a message where {A:B,C:D} is in the header, and {E:F} is not stored, but is included in the authentication.

For this message, go through the following test cases:
- customer supplies correct required key-value {E:F} (SUCCESS)
- customer supplies correct required key-value and additional correct header value {A:B,E:F} (SUCCESS)
- customer supplies does not supply required key-value {} (FAILURE)
- customer supplies incorrect value for required key-value {E:X} (FAILURE)
- customer supplies correct required key-value and additional incorrect header value {A:X,E:F} (FAILURE)
- customer supplies correct required key-value and additional incorrect, non-header key-value {E:F,X:Y} (FAILURE)

For every one of the above cases, attempt to decrypt using both a Default CMM, and a REC CMM with E configured as a "required key".

Next, suppose we have a message where {A:B,C:D,E:F} is in the header, and there is no additional non-stored key-values authenticated with the message.

For this message, go through the following test cases:
- customer supplies no reproduced context {} (SUCCESS)
- customer supplies additional correct header values {A:B,C:D,E:F} (SUCCESS)
- customer supplies incorrect value for key in header {A:X} (FAILURE)
- customer supplies incorrect, non-header key-value {X:Y} (FAILURE)

For each of these use a Default CMM.

---

This is more test cases than we currently have, and we may be able to convince ourselves that some of those test cases are not actually useful. However, I would prefer if this document focused on listing the parameters for the test cases in a way that is easy for readers to see whether we are missing any (And if we decide any are not useful, say why). If we do this, I don't think we need to give example code for each test case.